core/shared/src/main/scala/zio/stm/TArray.scala

```diff
@@ -16,139 +16,193 @@
-  final def exists(predicate: A => Boolean): STM[Nothing, Boolean] = find(predicate).map(_.isDefined)
+  /**
+   * Determine if the array contains a value satisfying a predicate.
+   */
+  final def exists(p: A => Boolean): STM[Nothing, Boolean] = find(p).map(_.isDefined)
 
-  /** Determine if the array contains a value satisfying an STM predicate. */
-  final def existsM[E](predicate: A => STM[E, Boolean]): STM[E, Boolean] =
-    countM(predicate).map(_ > 0) //Existence should not be order dependent so the entire array needs to be observed.
+  /**
+   * Determine if the array contains a value satisfying an STM predicate.
+   */
+  final def existsM[E](p: A => STM[E, Boolean]): STM[E, Boolean] =
+    countM(p).map(_ > 0)
 
-  /** Find the first element in the array matching a predicate. */
-  final def find(predicate: A => Boolean): STM[Nothing, Option[A]] =
+  /**
+   * Find the first element in the array matching a predicate.
+   */
+  final def find(p: A => Boolean): STM[Nothing, Option[A]] =
     if (array.isEmpty) STM.succeed(None)
     else
       array.head.get.flatMap { a =>
-        if (predicate(a)) STM.succeed(Some(a))
-        else new TArray(array.tail).find(predicate)
+        if (p(a)) STM.succeed(Some(a))
+        else new TArray(array.tail).find(p)
       }
 
-  /** Find the last element in the array matching a predicate. */
-  final def findLast(predicate: A => Boolean): STM[Nothing, Option[A]] =
-    new TArray(array.reverse).find(predicate)
+  /**
+   * Find the last element in the array matching a predicate.
+   */
+  final def findLast(p: A => Boolean): STM[Nothing, Option[A]] =
+    new TArray(array.reverse).find(p)
 
-  /** Find the last element in the array matching an STM predicate. */
-  final def findLastM[E](predicate: A => STM[E, Boolean]): STM[E, Option[A]] =
-    new TArray(array.reverse).findM(predicate)
+  /**
+   * Find the last element in the array matching an STM predicate.
+   */
+  final def findLastM[E](p: A => STM[E, Boolean]): STM[E, Option[A]] =
+    new TArray(array.reverse).findM(p)
 
-  /** Find the first element in the array matching an STM predicate. */
-  final def findM[E](predicate: A => STM[E, Boolean]): STM[E, Option[A]] =
+  /**
+   * Find the first element in the array matching an STM predicate.
+   */
+  final def findM[E](p: A => STM[E, Boolean]): STM[E, Option[A]] =
     if (array.isEmpty) STM.succeed(None)
     else
       array.head.get.flatMap { a =>
-        predicate(a).flatMap { result =>
+        p(a).flatMap { result =>
           if (result) STM.succeed(Some(a))
-          else new TArray(array.tail).findM(predicate)
+          else new TArray(array.tail).findM(p)
         }
       }
 
-  /** The first entry of the array, if it exists. */
-  final def firstOption: STM[Nothing, Option[A]] = if (array.isEmpty) STM.succeed(None) else array.head.get.map(Some(_))
+  /**
+   * The first entry of the array, if it exists.
+   */
+  final def firstOption: STM[Nothing, Option[A]] =
```

Comment:
Note that we could add a `first` that just retries if the array is empty 🤔

Reply:
Oh, I believe that was intentionally removed as we didn't want to expose "unsafe" methods (see #2446).

Reply:
Yes, my point is thanks to the semantics of STM, you can make a safe version of `first` that just parks the computation if the array is empty 😄 WDYT?

Reply:
True that. But the question is should we reflect that behavior to all accesses, e.g. non-existing index?

Reply:
hmm, good question. I wouldn't, e.g. `array(-1)` would never make sense.

Reply:
Ah, the thing is, the current API doesn't really allow elements to be added, so retrying won't make much sense.

Reply:
Ah right 👍